src/ui/public/visualize/visualize_legend.js

```diff
@@ -98,14 +104,28 @@ uiModules.get('kibana')
       ];
 
       function refresh() {
+        if (!$scope.renderbot) return;
         const vislibVis = $scope.renderbot.vislibVis;
 
         if ($scope.uiState.get('vis.legendOpen') == null && $scope.vis.params.addLegend != null) {
           $scope.open = $scope.vis.params.addLegend;
         }
 
-        $scope.labels = getLabels($scope.data, vislibVis.visConfigArgs.type);
-        $scope.getColor = colorPalette(_.pluck($scope.labels, 'label'), $scope.uiState.get('vis.colors'));
+        if (vislibVis.visConfigArgs.type === 'heatmap') {
```

Comment:
this should go away in the future, as all vislib charts will behave the same (expose the legend labels instead of legend going directly to the data)